### ISSUE TITLE
v0.11.4 config/env/staging.js Config Ignored

### ISSUE THREAD

**Issue by NikkiDreams** [balderdashy/sails#3544]
We have a multi-environment ecology with Staging thrown in there along with local, dev and prod. 
When we lift the server the staging config is read but the mysql adapter and connection and other config options are ignored. The default adapter which uses root@localhost is being pulled from, I assume, the sails-mysql module. It does not matter where we put the connections and model adapter config. It is never used. 

I have put both configs in their respective base config files and in the config/env/staging.js I even set log: 'silly' in the staging config file. It too is ignored. I have a console.log statement at the top of each config file. I see the correct files being read in based on each environment. But the staging config options are simply not used.

Same results occur no matter how we lift the server
```NODE_ENV=staging sails lift```
or 
```NODE_ENV=staging npm start```

Whether in config/... or config/env/... the options are not used. I confirmed this behavior on my own dev system as well as all of our remote servers and another developers dev system.
```
connections: {
    stagingMysqlServer: {
      adapter: 'sails-mysql',
      host: '<private ip>',
      user: 'root',
      password: 'xxxxx',
      database: 'mydb'
    }
  }

models: {
    connection: 'stagingMysqlServer'
}
```

**Comment by mikermcneil** [comment]
@NikkiDreams thanks for the report, we'll look into it asap.  In the mean time, would you provide your node and NPM versions as well?  (doubt it actually matters here, but we've been trying to be as disciplined as we can about getting that information documented in issues so that they're more useful when landed on in the future; e.g. from search results)  Thanks!

**Comment by sgress454** [comment]
@NikkiDreams do you have settings in your `local.js` file that may be overriding the environment config settings?  The `local.js` file is meant for a single system, thus [taking precedence over all other config](http://sailsjs.org/documentation/concepts/configuration#?the-config-localjs-file).

**Comment by NikkiDreams** [comment]
@sgress454 No I physically removed all other env/config files just to be sure that was not the case.

**Comment by sgress454** [comment]
@NikkiDreams this would be in `/config/local.js`, not under `/config/env`.

**Comment by NikkiDreams** [comment]
@sgress454 Yeah. I don't have a config/local.js at all. 
@mikermcneil Per your request. Sorry I should have put this info in there from the start.
 ```
npm version
{ 'rest-api': '0.2.0',
  npm: '3.5.3',
  ares: '1.10.1-DEV',
  http_parser: '2.5.0',
  icu: '56.1',
  modules: '46',
  node: '4.2.4',
  openssl: '1.0.2e',
  uv: '1.7.5',
  v8: '4.5.103.35',
  zlib: '1.2.8' }
```
and for totally nerding out details:
```
npm ls --depth=1
rest-api@0.2.0 /www/5m/rest-api
├── bluebird@3.1.5
├─┬ chai@3.5.0
│ ├── assertion-error@1.0.1
│ ├── deep-eql@0.1.3
│ └── type-detect@1.0.0
├─┬ debug@2.2.0
│ └── ms@0.7.1
├─┬ esdoc@0.4.3
│ ├── cheerio@0.19.0
│ ├── color-logger@0.0.3
│ ├── core-js@0.9.18
│ ├── escape-html@1.0.2
│ ├── escodegen@1.7.0
│ ├── espree@2.2.0
│ ├── estraverse@4.1.0
│ ├── fs-extra@0.20.1
│ ├── ice-cap@0.0.3
│ ├── marked@0.3.5
│ ├── minimist@1.1.1
│ └── taffydb@2.7.2
├── faker@3.0.1
├─┬ glob@6.0.4
│ ├── inflight@1.0.4
│ ├── inherits@2.0.1
│ ├── minimatch@3.0.0
│ ├── once@1.3.3
│ └── path-is-absolute@1.0.0
├── hoek@3.0.4
├─┬ istanbul@0.4.2
│ ├── abbrev@1.0.7
│ ├── async@1.5.2
│ ├── esprima@2.7.1
│ ├── fileset@0.2.1
│ ├── handlebars@4.0.5
│ ├── js-yaml@3.5.2
│ ├── mkdirp@0.5.1
│ ├── nopt@3.0.6
│ ├── once@1.3.3
│ ├── resolve@1.1.7
│ ├── supports-color@3.1.2
│ ├── which@1.2.4
│ └── wordwrap@1.0.0
├─┬ jsonapi-serializer@2.0.4
│ ├── inflection@1.8.0
│ └── lodash@3.10.1
├─┬ kerberos@0.0.18
│ └── nan@2.0.9
├── lodash@4.0.1
├─┬ mocha@2.4.5
│ ├── commander@2.3.0
│ ├── debug@2.2.0
│ ├── diff@1.4.0
│ ├── escape-string-regexp@1.0.2
│ ├── glob@3.2.3
│ ├── growl@1.8.1
│ ├── jade@0.26.3
│ ├── mkdirp@0.5.1
│ └── supports-color@1.2.0
├── UNMET DEPENDENCY newrelic@^1.25.1
├─┬ passport@0.3.2
│ ├── passport-strategy@1.0.0
│ └── pause@0.0.1
├─┬ passport-amazon-token@2.1.0
│ └── passport-oauth@1.0.0
├── passport-facebook-token@3.2.0
├── passport-foursquare-token@2.1.0
├── passport-github-token@2.1.0
├── passport-google-plus-token@2.1.0
├── passport-instagram-token@2.3.0
├─┬ passport-jwt@1.2.1
│ └── jsonwebtoken@5.5.4
├── passport-local@1.0.0
├── passport-paypal-token@2.1.0
├── passport-reddit-token@2.1.0
├── passport-soundcloud-token@2.1.0
├── passport-twitch-token@2.1.0
├── passport-twitter-token@1.1.0
├── passport-vkontakte-token@2.1.0
├── passport-windows-live-token@2.1.0
├── passport-yahoo-token@2.1.0
├── passport-yandex-token@2.1.0
├── pluralize@1.2.1
├── randomstring@1.1.3
├─┬ rc@1.1.6
│ ├── deep-extend@0.4.1
│ ├── ini@1.3.4
│ ├── minimist@1.2.0
│ └── strip-json-comments@1.0.4
├─┬ sails@0.11.4
│ ├── anchor@0.10.5
│ ├── async@0.9.2
│ ├── captains-log@0.11.11
│ ├── colors@0.6.2
│ ├── commander@2.1.0
│ ├── connect@2.30.0
│ ├── connect-flash@0.1.1
│ ├── cookie@0.1.2
│ ├── cookie-signature@1.0.6
│ ├── ejs@0.8.8
│ ├── ejs-locals@1.0.2
│ ├── express@3.21.2
│ ├── express-handlebars@2.0.1
│ ├── fs-extra@0.8.1
│ ├── glob@3.2.11
│ ├── grunt@0.4.2
│ ├── grunt-cli@0.1.13
│ ├── grunt-contrib-clean@0.5.0
│ ├── grunt-contrib-coffee@0.10.1
│ ├── grunt-contrib-concat@0.3.0
│ ├── grunt-contrib-copy@0.5.0
│ ├── grunt-contrib-cssmin@0.9.0
│ ├── grunt-contrib-jst@0.6.0
│ ├── grunt-contrib-less@0.11.1
│ ├── grunt-contrib-uglify@0.4.1
│ ├── grunt-contrib-watch@0.6.1
│ ├── grunt-sails-linker@0.9.6
│ ├── grunt-sync@0.0.8
│ ├── i18n@0.5.0
│ ├── include-all@0.1.6
│ ├── lodash@2.4.2
│ ├── merge-defaults@0.1.4
│ ├── method-override@2.3.5
│ ├── mock-req@0.1.0
│ ├── mock-res@0.1.0
│ ├── node-uuid@1.4.7
│ ├── pluralize@0.0.12
│ ├── prompt@0.2.14
│ ├── rc@0.5.5
│ ├── reportback@0.1.9
│ ├── sails-disk@0.10.8
│ ├── sails-generate@0.12.4
│ ├── sails-hook-sockets@0.12.3
│ ├── sails-stringfile@0.3.2
│ ├── sails-util@0.10.6
│ ├── semver@4.3.6
│ ├── skipper@0.5.9
│ ├── uid-safe@1.1.0
│ └── waterline@0.10.31
├─┬ sails-disk@0.10.8
│ ├── async@0.2.10
│ ├── fs-extra@0.8.1
│ ├── lodash@2.4.2
│ ├── waterline-criteria@0.11.2
│ ├── waterline-cursor@0.0.6
│ └── waterline-errors@0.10.1
├── sails-generate-jsonapi-blueprints@0.0.0 -> /www/sails-generate-jsonapi-blueprints extraneous
├─┬ sails-hook-autoreload@0.13.0
│ └── chokidar@1.4.2
├─┬ sails-hook-cron@2.2.0
│ └── cron@1.1.0
├─┬ sails-hook-dev@1.0.0
│ ├── fs-extra@0.16.5
│ ├── lodash@3.10.1
│ └── pretty-bytes@1.0.4
├─┬ sails-hook-jsonapi@0.0.1 -> /www/sails-hook-jsonapi invalid
│ ├── barrels@1.6.4
│ ├── eslint@1.10.3
│ ├── eslint-config-ianvs@0.0.6
│ ├── eslint-config-standard@4.4.0
│ ├── eslint-plugin-standard@1.3.1
│ ├── jsonapi-serializer@2.0.0-beta.1 (git://github.com/seyz/jsonapi-serializer.git#c3e8fb0788df18821c539cec053ac57211210ef3)
│ ├── pluralize@1.2.1
│ ├── sails-build-dictionary@0.10.1
│ ├── tap-spec@4.1.1
│ ├── tape@4.4.0
│ └── tape-catch@1.0.4
├─┬ sails-memory@0.10.5
│ ├── async@0.2.10
│ ├── lodash@2.4.2
│ └── waterline-criteria@0.10.7
├─┬ sails-mongo@0.12.0
│ ├── async@1.5.2
│ ├── lodash@3.10.1
│ ├── mongodb@2.1.3
│ └── validator@4.5.1
├─┬ sails-mysql@0.11.2
│ ├── async@1.3.0
│ ├── lodash@3.10.1
│ ├── mysql@2.8.0
│ └── waterline-sequel@0.5.5
├── sails-service-cipher@3.2.1
├─┬ sails-service-hash@3.2.1
│ └── bcryptjs@2.3.0
├─┬ sails-service-image@3.1.1
│ └── gm@1.21.1
├─┬ sails-service-location@3.3.1
│ └── node-geocoder@3.6.2
├─┬ sails-service-mailer@3.2.1
│ ├── nodemailer@2.0.0
│ ├── nodemailer-direct-transport@2.0.1
│ ├── nodemailer-sendgrid-transport@0.2.0
│ ├── nodemailer-sendmail-transport@1.0.0
│ ├── nodemailer-ses-transport@1.3.0
│ ├── nodemailer-smtp-transport@2.0.1
│ └── nodemailer-stub-transport@1.0.0
├─┬ sails-service-payment@3.1.1
│ ├── braintree@1.35.0
│ └── stripe@4.3.0
├─┬ sails-service-pusher@3.2.1
│ ├── apn@1.7.5
│ └── node-gcm@0.14.0
├─┬ sails-service-social@3.1.1
│ └── fb@1.0.2
├─┬ sails-service-storage@3.2.1
│ └── aws-sdk@2.2.32
└─┬ sinon@1.17.3
  ├── formatio@1.1.1
  ├── lolex@1.3.2
  ├── samsam@1.1.2
  └── util@0.10.3
```

**Comment by sgress454** [comment]
@NikkiDreams thanks.  Can you post your whole `config/env/staging.js` file too?  I'm not able to reproduce here, and our tests for env-specific config are passing, but I'm sure we'll get to the bottom of it.

**Comment by NikkiDreams** [comment]
@sgress454 
```
/**
 * Staging environment settings
 *
 * This file can include shared settings for a production environment,
 * such as API keys or remote database passwords.  If you're using
 * a version control solution for your Sails app, this file will
 * be committed to your repository unless you add it to your .gitignore
 * file.  If your repository will be publicly viewable, don't add
 * any private information to this file!
 *
 */
 console.log("___STAGE___");
module.exports =  {

  /***************************************************************************
   * Set the default database connection for models in the production        *
   * environment (see config/connections.js and config/models.js )           *
   ***************************************************************************/

  // models: {
  //   connection: 'someMysqlServer'
  // },

  /***************************************************************************
   * Set the port in the production environment to 80                        *
   ***************************************************************************/

  // port: 80,

  /***************************************************************************
   * Set the log level in production environment to "silent"                 *
   ***************************************************************************/

  connections: {
    stagingMysqlServer: {
      adapter: 'sails-mysql',
      host: '123.123.123.123',
      user: 'me',
      password: 'xxxxxxxx',
      database: 'xxxxxxxx'
    }
  },

  local: {
    port: 3001,
    environment: 'staging'
  },

  blueprints: {
    prefix: '/api',
    restPrefix: '/v1',
    jsonapi: {
      sideload: true
    }
  },

  models: {
    connection: 'stagingMysqlServer',
    migrate: 'alter',
    autoCreatedAt: true,
    autoUpdatedAt: true,
    autoPK: true,
    schema: false,
    associations: {
      list: "link",
      detail: "record"
    },
    validations: {
      ignoreProperties: [ 'includeIn' ]
    }
  },

  log: {
    level: 'error'
  },

  orm:{
    _hookTimeout: 45000
  }
};

```

**Comment by sgress454** [comment]
@NikkiDreams Thanks.  One question I have is, what config (if any) _is_ being loaded?  You might try starting your app with `NODE_ENV=staging sails console` and then examining various `sails.config` properties (for example by typing `sails.config.connections`).  If you see something in there that shouldn't be, you then search your app code to determine where it might be coming from.

For my part, here's what I did to try and reproduce this issue:

1. Create a new Sails app with `sails new` using Sails v0.11.4
1. `npm install sails-mysql`
1. Created a `config/env/staging.js` file using the contents you provided
1. Change the MySQL config to connect to a local MySQL server and database
1. Removed all other config files

With these steps, the config is definitely being loaded for me.  I can confirm by doing `NODE_ENV=staging sails console` and examining various `sails.config` properties like `sails.config.connections`.

If `sails console` with your app doesn't reveal anything, would you mind trying a similar test with a fresh app?

**Comment by NikkiDreams** [comment]
@sgress454 I will give it a whack and get back to you. 
BTW thank you so much for the very quick responses. I know you guys are busy too. :-)

**Comment by NikkiDreams** [comment]
@sgress454
Well I am happy to say things are working now. 
This was so strange. Running sails console and sails.config showed me that absolutely none of the file was being loaded/read past the first block of comments above the module.export = {}

I did not know about using sails console. So happy you told me about that one, because I would never have figured this out in such a short time. 

**## Thank you.**

FYI: I removed all comments, JSlint'd and Pretty printed the config/env/staging.js file and everything started working correctly. That file was literally a copy of the config/env/production.js file. The only thing I can think of that would cause such a thing might be the header BOMB in the file itself or one of those mysterious, invisible chars that do not show up in most all editors.
